tutorials/hpc_clusters.md

```diff
@@ -0,0 +1,406 @@
+      rendered by the GPU. This will work with all further described
+      approaches, except approach 3 (`--headless-rendering`).
+      To use `glxgears`, you need to install package `mesa-utils` (on
+      Debian-based systems) or `glx-utils` (on Fedora-based systems) inside
+      your image. Then, instead of `gz sim -v4 -r -s sensors_demo.sdf`, type
+      `glxgears -info | cut -c -80` and look at the value of `GL_RENDERER`.
+      `llvmpipe` is software rendering (slow), anything else should be GPU-accelerated.
+
+\note The approaches are ordered by the simplicity of their use/setup if they
+      work. However, most HPC clusters will probably not support
+      Approach 1 and 2, so you can also start with Approach 3 first and then go
+      to 1 and 2 if it doesn't work.
+
+## Approach 1: Initial test
+
+\note **Do not try to run GUI on the cluster unless the server-only part
+      works**. It would only complicate things. To run only Gazebo server, make
+      sure flag `-s` is in the `gz sim` command line (and `-g` is not).
+
+*This tests Gazebo in GLX mode connected to an existing X server.*
+
+Try directly running Gazebo on the GPU node. For easier testing and to rule-out
+ROS complexity, use the `sensors_demo.sdf` world that comes preinstalled with
+Gazebo. If you don't have it, just
+[download it](https://github.com/gazebosim/gz-sim/blob/main/examples/worlds/sensors_demo.sdf)
+e.g. to your home folder and provide an absolute path to the file instead of
+simply `sensors_demo.sdf`.
+
+```bash
+user@cluster$ srun -p gpufast--gres=gpu:1 --pty bash -i
+user@gpu-node-1$ singularity exec image.sif bash -c "echo $DISPLAY; gz sim -v4 -s -r sensors_demo.sdf & (gz topic -t /camera -e | cut -c -80) & sleep 30; kill %1; kill %2; sleep 2; kill -9 %1; kill -9 %2"
+```
+
+This tests Gazebo on the interactive GPU partition. Now, let's do the same test
+on a noninteractive one. Prepare the following job spec into file `test.batch`:
+
+```bash
+#!/bin/sh
+#SBATCH --time=1 -p gpu --gres=gpu:1
+singularity exec image.sif bash -c "gz sim -v4 -s -r sensors_demo.sdf & (gz topic -t /camera -e | cut -c -80) & sleep 30; kill %1; kill %2; sleep 2; kill -9 %1; kill -9 %2"
+```
+
+Of course, replace `-p gpu` with the name of your non-interactive GPU
+partition.
+
+Schedule the batch for execution and wait until it is executed. Check
+`test.stdout` file for the output.
+
+```bash
+user@cluster$ sbatch -n1 -o test.stdout test.batch
+user@cluster$ tail -F test.stdout
+```
+
+If the commands did not fail, did not print anything red and you see camera
+messages printed out, rendering is working for you. Now it's time to find out
+if it is HW-accelerated:
+
+```bash
+user@cluster$ grep -C2 RENDERER ~/.gz/rendering/ogre2.log
+```
+
+If you see `llvmpipe` in the output, the rendering is not HW-accelerated and
+you should keep following this guide (unless you're happy with slower
+rendering). If the output contains `NVidia` (or `AMD`), tadaa, you have a
```

Comment:
sample output below. The nvidia string may be in the `GL_VENDOR` field.
```
23:39:53: GL_VERSION = 4.5.0 NVIDIA 435.21
23:39:53: GL_VENDOR = NVIDIA Corporation
23:39:53: GL_RENDERER = GeForce GTX 1080/PCIe/SSE2
```

Reply:
Thanks, I know about this. That's why the command is `grep -C2` to provide a bit of context around the RENDERER line.